test/boost/expr_test.cc

```diff
@@ -1303,4 +1303,56 @@ BOOST_AUTO_TEST_CASE(evaluate_bind_variable_validates_empty_value_in_maps_recurs
 
     constant map_with_empty_value2 = create_nested_map_with_value(make_int_const(1), make_empty_const(int32_type));
     check_bind_variable_evaluate(map_with_empty_value2, expected_valid);
-}
\ No newline at end of file
+}
+
```

Comment:
> It's interesting that prepare_expression
for column identifiers doesn't require a receiver.
I hope this won't break validation in the future.

prepare_expression has four modes:
 - with a receiver 
     - with the prepared expression having a known type
        check that the types are compatible
     - with the prepared expression not having a known type (like bind variables)
        infer the prepared type from the reciever
 - without a receiver
     - with the prepared expression having a known type
        accept the prepared expression type
     - with the prepared expression not having a known type (like bind variables)
        reject the expression

For any given expression just two are applicable, and we should test both of them. What you have now is okay to start with.